### PR TITLE
fix(syscall): handle invalid input in bn256 syscalls

### PR DESCRIPTION
**Pull request by magicalne**
Handle errors in bn256 and return errors for syscalls.

### CODE REVIEW

File: crates/generator/src/syscalls/bn.rs [hunk 1]
```diff
@@ -1,27 +1,36 @@
...
-fn read_pt(buf: &[u8]) -> Result<G1, Error> {
-    let px = Fq::from_slice(&buf[0..32]).map_err(|_| Error("invalid pt"))?;
-    let py = Fq::from_slice(&buf[32..64]).map_err(|_| Error("invalid pt"))?;
+type Result<T> = std::result::Result<T, BnError>;
+fn read_pt(buf: &[u8]) -> Result<G1> {
+    let map_err = BnError::G1FieldError;
+    let px = Fq::from_slice(&buf[0..32]).map_err(map_err)?;
+    let py = Fq::from_slice(&buf[32..64]).map_err(map_err)?;
     Ok(if px == Fq::zero() && py == Fq::zero() {
         G1::zero()
```
Comment (jjyr) [inline crates/generator/src/syscalls/bn.rs:23]: Is this a breaking change? (is `AffineG1(Fq::zero(), Fq::zero()) == G1::zero()`? )

**Review by jjyr — DISMISSED** [review]

File: crates/generator/src/syscalls/bn.rs (hunk 1 above)
Comment (magicalne) [reply crates/generator/src/syscalls/bn.rs:23]: Well, actually I don't break any change here. But after checking the code, looks like  `AffineG1(Fq::zero(), Fq::zero())` and `G1::zero()` are not equivalent.
`G1::zero()` represents:
```rust
        G {
            x: P::Base::zero(),
            y: P::Base::one(),
            z: P::Base::zero(),
        }
```

And for `AffineG1(Fq::zero(), Fq::zero())`:
```rust
      G {
           x: x,
           y: y,
           z: P::Base::one(),
     };
```

@Flouse Do you remember about this? Is that intended?

File: crates/generator/src/syscalls/mod.rs [hunk 2]
```diff
@@ -565,29 +569,39 @@ impl<'a, 'b, S: State + CodeStore + JournalDB, C: ChainView, Mac: SupportMachine
...
-                    log::error!("{}", err_msg);
-                    VMError::Unexpected(err_msg)
-                })?;
-                store_data(machine, output.as_slice())?;
-
-                machine.set_register(A0, Mac::REG::from_u8(SUCCESS));
+                log::debug!("[bn256 input]: {:?}", &input[0..32]);
+                log::debug!("[bn256 input]: {:?}", &input[32..64]);
+                log::debug!("[bn256 input]: {:?}", &input[64..96]);
+                log::debug!("[bn256 input]: {:?}", &input[96..128]);
```
Comment (Flouse) [inline crates/generator/src/syscalls/mod.rs]: The `input_size` can be less than 128. This debug log may cause panic.

**Review by Flouse — CHANGES_REQUESTED** [review]

File: crates/generator/src/syscalls/bn.rs (hunk 1 above)
Comment (Flouse) [reply crates/generator/src/syscalls/bn.rs:23]: Yes, that is intended.

As my understanding, when `px == Fq::zero() && py == Fq::zero()`,
it is a special coordinate at infinity with `z == P::Base::zero()`. 

similar implementation: https://github.com/openethereum/parity-ethereum/blob/c7f608ec740882eac94038249037ddd955b60d31/ethcore/src/builtin.rs#L408-L414

**Review by Flouse — DISMISSED** [review]

**Review by Flouse — APPROVED** [review]

**Review by jjyr — APPROVED** [review]

**Pull request merged**